openwillis/measures/text/util/speech/coherence.py

```diff
@@ -0,0 +1,557 @@
+        # calculate pseudo-perplexity of the turn and indicating how predictable the turn is
+        perplexity, perplexity_5, perplexity_11, perplexity_15 = calculate_perplexity(utterance_text, bert, tokenizer)        
+
+    return sentence_tangeniality1, sentence_tangeniality2, perplexity, perplexity_5, perplexity_11, perplexity_15
+
+def calculate_slope(y):
+    """
+    ------------------------------------------------------------------------------------------------------
+    This function calculates the slope
+     of the input list using linear regression
+
+    Parameters:
+    ...........
+    y: list
+        A list of values.
+
+    Returns:
+    ...........
+    float
+        The calculated slope of the input list.
+
+    ------------------------------------------------------------------------------------------------------
+    """
+    # remove NaNs
+    y = [val for val in y if not np.isnan(val)]
+
+    x = range(len(y))
+    try:
+        slope, _ = np.polyfit(x, y, 1)
+    except:
+        slope = np.nan
+
+    return slope
+
+def init_model(language, measures):
+    """
+    ------------------------------------------------------------------------------------------------------
+
+    This function initializes the appropriate models and tokenizers based on language.
+
+    Parameters:
+    ...........
+    language: str
+        Language of the transcribed text.
+    measures: dict
+        A dictionary containing the names of the columns in the output dataframes.
+
+    Returns:
+    ...........
+    sentence_encoder: SentenceTransformer
+        A SentenceTransformer model.
+    tokenizer: BertTokenizer
+        A BERT tokenizer.
+    bert: BertForMaskedLM
+        A BERT model.
+
+    ------------------------------------------------------------------------------------------------------
+    """
+    sentence_encoder, tokenizer, bert = None, None, None
+    if language in measures["english_langs"]:
+        sentence_encoder = SentenceTransformer('sentence-transformers/all-MiniLM-L6-v2')
+
+        tokenizer = BertTokenizer.from_pretrained('bert-base-cased')
+        bert = BertForMaskedLM.from_pretrained('bert-base-cased')
+    else:
+        if language in measures["supported_langs_sentence_embeddings"]:
+            sentence_encoder = SentenceTransformer('sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2')
+        else:
+            logger.error(f"Language {language} not supported for phrase coherence analysis")
+
+        if language in measures["supported_langs_bert"]:
+            tokenizer = BertTokenizer.from_pretrained('bert-base-multilingual-cased')
+            bert = BertForMaskedLM.from_pretrained('bert-base-multilingual-cased')
+        else:
+            logger.error(f"Language {language} not supported for perplexity analysis")
+
+    return sentence_encoder, tokenizer, bert
+
+def get_phrase_coherence(df_list, utterances_filtered, min_coherence_turn_length, speaker_label, language, measures):
```

Comment:
split this function into multiple helper functions

Reply:
addressed